packages/edit-site/src/components/global-styles/hooks.js

```diff
@@ -0,0 +1,347 @@
+			const currentConfig = parseContent(
+				getEditedEntityRecord(
+					'postType',
+					'wp_global_styles',
+					globalStylesId
+				)?.content
+			);
+			const updatedConfig = callback( currentConfig );
+			editEntityRecord( 'postType', 'wp_global_styles', globalStylesId, {
+				content: JSON.stringify( {
+					...updatedConfig,
+					settings: removeUserOriginFromSettings(
+						updatedConfig.settings
+					),
+				} ),
+			} );
+		},
+		[ globalStylesId ]
+	);
+
+	return [ config, setConfig ];
+}
+
+function useGlobalStylesBaseConfig() {
+	const baseConfig = useSelect( ( select ) => {
+		return select( editSiteStore ).getSettings()
+			.__experimentalGlobalStylesBaseStyles;
+	}, [] );
+
+	return baseConfig;
+}
+
+export function useGlobalStylesConfig() {
+	const [ userConfig, setUserConfig ] = useGlobalStylesUserConfig();
+	const baseConfig = useGlobalStylesBaseConfig();
+	const mergedConfig = useMemo( () => {
+		return mergeBaseAndUserConfigs( baseConfig, userConfig );
+	}, [ userConfig, baseConfig ] );
+
+	return [ baseConfig, userConfig, mergedConfig, setUserConfig ];
+}
+
+export const useGlobalStylesReset = () => {
+	const [ config, setConfig ] = useGlobalStylesUserConfig();
+	const canReset = !! config && ! isEqual( config, EMPTY_CONFIG );
+	return [
+		canReset,
+		useCallback( () => setConfig( () => EMPTY_CONFIG ), [ setConfig ] ),
+	];
+};
+
+export function useSetting( path, blockName, source = 'all' ) {
+	const [
+		baseConfig,
+		userConfig,
+		mergedConfig,
+		setUserConfig,
+	] = useGlobalStylesConfig();
+
+	const fullPath = ! blockName
+		? `settings.${ path }`
+		: `settings.blocks.${ blockName }.${ path }`;
+
+	const setSetting = ( newValue ) => {
+		setUserConfig( ( currentConfig ) => {
+			const newUserConfig = cloneDeep( currentConfig );
+			const pathToSet = PATHS_WITH_MERGE[ path ]
+				? fullPath + '.user'
+				: fullPath;
+			set( newUserConfig, pathToSet, newValue );
+
+			return newUserConfig;
+		} );
+	};
+
+	const getSettingValueForContext = ( name ) => {
+		const currentPath = ! name
+			? `settings.${ path }`
+			: `settings.blocks.${ name }.${ path }`;
+
+		const getSettingValue = ( configToUse ) => {
+			const result = get( configToUse, currentPath );
+			if ( PATHS_WITH_MERGE[ path ] ) {
+				return result?.user ?? result?.theme ?? result?.core;
+			}
+			return result;
+		};
+
+		let result;
+		switch ( source ) {
+			case 'all':
+				result = getSettingValue( mergedConfig );
+				break;
+			case 'user':
+				result = getSettingValue( userConfig );
+				break;
+			case 'base':
+				result = getSettingValue( baseConfig );
+				break;
+			default:
+				throw 'Unsupported source';
+		}
+
+		return result;
+	};
+
+	// Unlike styles settings get inherited from top level settings.
+	const resultWithFallback =
+		getSettingValueForContext( blockName ) ?? getSettingValueForContext();
+
+	return [ resultWithFallback, setSetting ];
+}
+
+export function useStyle( path, blockName, source = 'all' ) {
```

Comment:
Yeah, I know. This PR shouldn't be blocked by this. it's something we can iterate on. Though my concern with that approach is that it ties consumers to a specific `theme.json` shape, which we tried to avoid by separating "path" and "context" in the hooks/utilities.